### DISCUSSION TITLE
Ijwhost.dll location?

### DISCUSSION THREAD

**Discussion by berrywing012** [cefsharp/CefSharp#3440] [Q&A]
**_Where is the Ijwhost.dll file?_**

Everything works great in debug in a C# Windows Forms application.
However using the Setup MSI Wizard we discovered that a required file: Ijwhost.dll is missing.
The application installed with the MSI on other Windows machines will not run. 

Please note: This question is from a C# developer, not an assembly or C++ coder.
This was helpful, but the remaining file we need is the ljwhost.dll
https://github.com/cefsharp/CefSharp/wiki/Output-files-description-table-%28Redistribution%29

**Comment by amaitland** [comment]
What is your question exactly?

**Reply by berrywing012** [reply]
Thank you in advance. 
We are missing a dependent file for the installation package. The file Ijwhost.dll
Where is this file?
What is the path and how is it created?
Is it part of the GitHub library that is not available?

Using C#
We have a working C# project executable with a ChromiumWebBrowser.
Added a Visual Studio Installer project using the Setup Wizard for our C# application.
Not all dependencies are found.
We added all the redistribution files in this article.
https://github.com/cefsharp/CefSharp/wiki/Output-files-description-table-%28Redistribution%29

However Ijwhost.dll is listed as required and we do not know where it is. 
How do we get the redistribution / MSI to work with the CefSharp library?

**Reply by amaitland** [reply]
> Where is this file?

What do you mean? The file is included in the Nuget package. 

> What is the path and how is it created?
Is it part of the GitHub library that is not available?

It's part of the .Net SDK . There isn't a lot of documentation from Microsoft that I'm aware of. https://github.com/dotnet/runtime/issues/38231 has the best description I've seen.

> However Ijwhost.dll is listed as required and we do not know where it is.

It should have been copied to your projects bin folder like the rest of the dependencies. Your app wouldn't run without it.

Just include the file in your MSI like any other.

For reference 
https://github.com/cefsharp/CefSharp/blob/cefsharp/88/NuGet/PackageReference/CefSharp.Common.NETCore.nuspec#L42
https://github.com/cefsharp/CefSharp/blob/cefsharp/88/NuGet/PackageReference/CefSharp.Common.NETCore.nuspec#L46

I'm still not clear on what the problem is.

**Reply by amaitland** [reply]
If you'd like to know the advanced technical details then read the following 

https://github.com/dotnet/runtime/blob/main/docs/design/features/IJW-activation.md
https://github.com/dotnet/runtime/blob/main/docs/design/features/native-hosting.md

**Reply by berrywing012** [reply]
Sorry, I hope we are not coming across as difficult. 

Honestly, we cannot find this DLL anywhere, hence the confusion. 
The file Ijwhost.dll does not exist in any project path. 
It is not found in any searches or directories mentioned in articles above.
If it does exist, what is the path we should be looking at?

These locations have turned up empty.
..\..\packages\CefSharp.Common.88.2.90\CefSharp\x86\
..\..\CefSharp.Core.Runtime\bin.netcore\x64\Release\Ijwhost.dll

Even a search of the whole hard drive has turned up empty.
![Ijwhost](https://user-images.githubusercontent.com/2632504/110708763-c98b8500-81c0-11eb-9ffa-1e87b42add66.png)

**Reply by amaitland** [reply]
What is the exact package are you using? 

The CefSharp.Common 88.2.90 package targets .net 452 which doesn't require or include this file.

> For CefSharp.WinForms.NETCore/CefSharp.Wpf.NETCore/CefSharp.OffScreen.NETCore packages this file is required
To support C++/CLI libraries in .NET Core, ijwhost was created as a shim for finding and loading the runtime. All C++/CLI libraries are linked to this shim, such that ijwhost.dll is found/loaded when the C++/CLI library is loaded.

As per https://github.com/cefsharp/CefSharp/wiki/Output-files-description-table-%28Redistribution%29

Only the NETCore packages require ijwhost.dll

> ...\CefSharp.Core.Runtime\bin.netcore\x64\Release\Ijwhost.dll

That's The path of the file if you compiled the project from source which is then included in the Nuget package.

**Reply by berrywing012** [reply]
Oh, did we use the wrong NuGet package? 
The one we selected is working just fine.
![cefsharp1](https://user-images.githubusercontent.com/2632504/110711513-4caeda00-81c5-11eb-818d-ea410605e102.png)

**Reply by amaitland** [reply]
> Oh, did we use the wrong NuGet package?

I don't know what framework version you are targeting. 

As you have CefSharp.Winforms installed and not CefSharp.Winforms.NETCore you don't need ijwhost.dll

How did you come to the conclusion you needed ijwhost.dll?

**Reply by berrywing012** [reply]
How did you come to the conclusion you needed ijwhost.dll?
We assumed it is the problem. The application will not run after installing via MSI on another PC. 
As per https://github.com/cefsharp/CefSharp/wiki/Output-files-description-table-%28Redistribution%29
It said it was required so we tried to add it as we thought that could be the problem why it does not run  on PCs other than development.

**Reply by amaitland** [reply]
As per the description only the NETCore packages require ijwhost.dll. 

Make sure your target machine has Visual C++ installed. 

https://github.com/cefsharp/CefSharp/wiki/Output-files-description-table-(Redistribution)#visual-c
